pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -63,7 +60,7 @@ def __init__(self, min_steps: int, max_steps: int):
     @property
     def batch_idx(self) -> int:
         """Returns the current batch index (within this epoch)"""
-        return self.iteration_count
+        return self.batch_progress.current.ready - 1
```

Comment:
Let's have a test to make sure ready is always at the right place and batch_idx is correctly computed.

Reply:
Why not use completed there ?

Reply:
> Let's have a test to make sure ready is always at the right place and batch_idx is correctly computed.

We already have good tests for the progress properties in `test_loops`

> Why not use completed there ?

Because if a piece of code calls this property between `completed++` and the next `ready++`, the `batch_idx` will be one more than it should.

But I haven't observed this happen, so maybe I am overthinking it.

Reply:
Ok, makes sense.

Reply:
```suggestion
        # use `ready` instead of `completed` in case this is accessed after `completed` has been increased
        # but before the next `ready` increase
        return self.batch_progress.current.ready - 1
```